### PR TITLE
[REVIEW] Replace index type-dispatch call with indexalator in cudf::gather

### PR DESCRIPTION
**Pull request by davidwendt**
The `cudf::gather` API performs a double type-dispatch. The first one is against the index column parameter which will have an integer type like `int8, int16, int32, int64, uint8, uint16, uint32, or uint64`. The second dispatch is for the gather column data itself. This means the compiler will generate 8 copies of the gather code (second dispatch) where the only difference is the memory access to the index data. 

Using the index normalizing iterator (indexalator) introduced to help with dictionary support for the first dispatch greatly reduces the code size and compile time for `gather.cu`. The compile time was reduced on my system from 200s to 40s. The object file size changed from 10.8MB to 4.2MB. This also resulted in a size decrease for `libcudf.so` of 4MB.

Also, running the `gbenchmark/GATHER_BENCH` showed no significant change in performance using the indexalator over the type-dispatch. The following is the result of comparing `GATHER_BENCH` before and after replacing the first type-dispatch with the indexalator.
gather-benchmark
The chart shows the difference in values in the outputs. The negative values indicate the indexalator output was faster/better while the positive ones were slower/worse.

### CODE REVIEW

File: cpp/src/copying/gather.cu [hunk 1]
```diff
@@ -74,17 +39,37 @@ std::unique_ptr<table> gather(table_view const& source_table,
...
-                                                                   dispatch_map_type{},
-                                                                   source_table,
-                                                                   gather_map,
-                                                                   gather_map.size(),
-                                                                   bounds,
-                                                                   neg_indices,
-                                                                   mr,
-                                                                   stream);
+  // create index type normalizing iterator for the gather_map
+  auto map_begin = indexalator_factory::make_input_iterator(gather_map);
```
Comment (jrhemstad) [inline cpp/src/copying/gather.cu:43]: Looking at the indexalator factory functions, there should be a check to require that the input column does not contain nulls.

Comment (davidwendt) [reply cpp/src/copying/gather.cu:43]: This makes sense. I have use cases for dictionary where I want to do operations on dictionary indices along with a null mask  generally using pair iterators. But to build a pair iterator requires sometimes allowing the indexalator to be created with a nullable column.

Comment (jrhemstad) [reply cpp/src/copying/gather.cu:43]: > But to build a pair iterator requires sometimes allowing the indexalator to be created with a nullable column.

Sure, but this should use a private code path. The public factories should disallow columns with nulls.

Comment (davidwendt) [reply cpp/src/copying/gather.cu:43]: OK. Adding the exception will not effect this PR but will break some code unrelated to this PR so I would rather make this change in a separate PR. In that PR I would want to move the current code logic that creates a pair-indexalator to a utility or the indexalator factory too.
https://github.com/rapidsai/cudf/blob/48e9d1bb15c73f241a173bb6de02dbf496aa87bf/cpp/src/dictionary/replace.cu#L72-L77

Comment (davidwendt) [reply cpp/src/copying/gather.cu:43]: Created #6469 to track this

File: cpp/src/copying/gather.cu [hunk 2]
```diff
@@ -74,17 +39,37 @@ std::unique_ptr<table> gather(table_view const& source_table,
...
+  auto map_end   = map_begin + gather_map.size();
 
-  return destination_table;
+  if (bounds == out_of_bounds_policy::FAIL) {
+    cudf::size_type begin =
+      neg_indices == negative_index_policy::ALLOWED ? -source_table.num_rows() : 0;
+    cudf::size_type end = source_table.num_rows();
+    CUDF_EXPECTS(gather_map.size() == thrust::count_if(rmm::exec_policy()->on(0),
```
Comment (harrism) [inline cpp/src/copying/gather.cu:50]: Yikes!
```suggestion
    CUDF_EXPECTS(gather_map.size() == thrust::count_if(rmm::exec_policy(stream)->on(stream),
```

Comment (davidwendt) [reply cpp/src/copying/gather.cu]: Good catch. This has been in there forever. Also, this file was also missing a copyright header.

**Review by jrhemstad — APPROVED** [review]

**Review by rgsl888prabhu — APPROVED** [review]

**Pull request merged**